### PR TITLE
Token Rotation Improvements

### LINKED ISSUE #101: Invalidating previous access tokens during rotation
§6.1 Rotating the Access Token: Editor's note: 

> If the token is validated and the key is appropriate for the request, the AS MUST invalidate the current access token associated with this URL, if possible, and return a new access token response as described in Section 3.2.1, unless the multi_token flag is specified in the request. 

We could also use different verbs to signal whether or not the old token should be kept around or not, instead of using a token flag to do this.

### LINKED ISSUE #102: Require inclusion of token management URI in rotation response
§6.1 Rotation the Access Token: Editor's note: 

The net result is that the client's always going to use the management URL that comes back. But should we let the server omit it from the response if it doesn't change? That seems like an odd optimization that doesn't help the client.

### PR DESCRIPTION
**Pull request by jricher**
- update token rotation response and requirements
- add considerations for the difficulties of rotating and revoking stateless tokens
- add considerations on rotation in unreliable networks

Closes #101
Closes #102

### CODE REVIEW

File: draft-ietf-gnap-core-protocol.md [hunk 1]
```diff
@@ -3321,25 +3327,31 @@ The AS validates that the token presented is associated with the management
 URL, that the AS issued the token to the given client instance, and that
 the presented key is appropriate to the token.
 
-If the access token has expired, the AS SHOULD honor the rotation request to
-the token management URL since it is likely that the client instance is attempting to
+Note that in many cases, the access token will have expired for regular use. To facilitate
+token rotation, the AS SHOULD honor the rotation request of the expired access token
+since it is likely that the client instance is attempting to
 refresh the expired token. To support this, the AS MAY apply different lifetimes for
```
Comment (fimbault) [inline draft-ietf-gnap-core-protocol.md:3333]: actually, what we're saying here is that the lifetime for token management might be longer that for its use

File: draft-ietf-gnap-core-protocol.md [hunk 2]
```diff
@@ -3321,25 +3327,31 @@ The AS validates that the token presented is associated with the management
...
 the use of the token in management vs. its use at an RS. An AS MUST NOT
-honor a rotation request for an access token that has been revoked, either by
-the AS or by the client instance through the [token management URI](#revoke-access-token).
+honor a rotation request for an access token that has been revoked or otherwise disabled.
```
Comment (fimbault) [inline draft-ietf-gnap-core-protocol.md:3335]: what do we mean by disabled ?

File: draft-ietf-gnap-core-protocol.md [hunk 3]
```diff
@@ -3409,6 +3422,8 @@ means, the AS SHOULD honor the revocation request to
 the token management URL as valid, since the end result is still the token
 not being usable.
 
+Note that some deployment decisions like self-contained stateless access tokens can make
```
Comment (fimbault) [inline draft-ietf-gnap-core-protocol.md:3425]: isn't it a bit redundant with what is already said above ? Note that stateless access tokens can make proactive
revocation difficult within a system, see {{security-stateless-tokens}}.

File: draft-ietf-gnap-core-protocol.md [hunk 4]
```diff
@@ -5354,6 +5369,61 @@ through [discovering the AS from the RS](#rs-request-without-token).
...
+should probably not offer token revocation functionality to client instances, since the client
+instance's calls to such an endpoint is effectively meaningless. However, a client instance calling
+the token revocation function will also throw out its copy of the token, so such a placebo endpoint
+might not be completely meaningless. Token rotation similarly difficult because the AS has to
+revoke the old access token after a rotation call has been made. If the access tokens are
+completely self-contained and non-revocable, this means that there will be a period of time during
+which both the old and new access tokens are valid and usable, which is an increased security risk
+for the environment.
+
+These problems can be mitigated by keeping the validity time windows of self-contained access tokens
```
Comment (fimbault) [inline draft-ietf-gnap-core-protocol.md:5394]: probably we should say that the mitigations should be applied for specific use cases, that aren't too sensitive

**Review by fimbault — COMMENTED** [review]
a few minor comments

File: draft-ietf-gnap-core-protocol.md (hunk 2 above)
Comment (jricher) [reply draft-ietf-gnap-core-protocol.md:3335]: Generically, just "made unusable for any reason whatsoever". Revocation is the most concrete -- and permanent -- version of that.

File: draft-ietf-gnap-core-protocol.md (hunk 3 above)
Comment (jricher) [reply draft-ietf-gnap-core-protocol.md:3425]: I believe you're correct, this paragraph can be dropped.

File: draft-ietf-gnap-core-protocol.md [hunk 5]
```diff
@@ -3409,6 +3422,8 @@ means, the AS SHOULD honor the revocation request to
 the token management URL as valid, since the end result is still the token
 not being usable.
 
+Note that some deployment decisions like self-contained stateless access tokens can make
+propagating revocation difficult, see {{security-stateless-tokens}}.
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:3426]:
```suggestion
```

File: draft-ietf-gnap-core-protocol.md [hunk 6]
```diff
@@ -3321,25 +3327,31 @@ The AS validates that the token presented is associated with the management
 URL, that the AS issued the token to the given client instance, and that
 the presented key is appropriate to the token.
 
-If the access token has expired, the AS SHOULD honor the rotation request to
-the token management URL since it is likely that the client instance is attempting to
+Note that in many cases, the access token will have expired for regular use. To facilitate
+token rotation, the AS SHOULD honor the rotation request of the expired access token
+since it is likely that the client instance is attempting to
 refresh the expired token. To support this, the AS MAY apply different lifetimes for
 the use of the token in management vs. its use at an RS. An AS MUST NOT
```
Comment (jricher) [inline draft-ietf-gnap-core-protocol.md:3334]:
```suggestion
refresh the expired token. To support this, the AS MAY allow a longer lifetime for
token management compared to its use at an RS. An AS MUST NOT
```

**Pull request merged**